### ISSUE TITLE
Skip link doesn't become visible when receiving focus from VoiceOver on iOS or iPadOS 

### ISSUE THREAD

**Issue by dav-idc** [alphagov/govuk-frontend#4098]
## Overview
During [manual accessibility testing for the Exit this Page component](https://docs.google.com/spreadsheets/d/1ruOHpJ4ERL8nshwLpQ7N6WGoFscrFwkr215gxsVPO-Q/edit?usp=sharing), we uncovered two potential bugs related to the 'skip link'.

The plan is to investigate these issues at the 'skip link' level, and adopt any changes made into the Exit this Page secondary link.

Issue 1 is covered by this GitHub issue, and issue 2 is covered in the following issue:
- https://github.com/alphagov/govuk-frontend/issues/4100

## Description of the issue
When navigating using VoiceOver in Safari on iOS or iPadOS, the skip link doesn't appear when the screen reader focus moves to the link. The skip link is announced, but isn't a visible element. The skip link does become visible on screen reader focus for other screen readers.

## Steps to reproduce the issue
1. Turn on VoiceOver on iOS or iPadOS first
2. Open the GOV.UK Design System website in Safari
3. From the top of the page, press 'tab' once
4. Check the following:
    - does the screen reader successfully announce the skip link? (we expect yes)
    - does the skip link appear on the screen? (we expect no, but ideally yes)

## Actual vs expected behaviour

We would expect the skip link to become visible when the screen reader focus moves to the skip link.
Instead, the skip link is not visibly shown by VoiceOver on focus.

Here's a screenshot of:
- the [GOV.UK Design System homepage](https://design-system.service.gov.uk/)
- on iOS 16, iPhone 12 Pro
- with VoiceOver turned on and focussed on the hidden skip link
<img src="https://github.com/alphagov/govuk-frontend/assets/98318778/124128dc-7dae-4744-b505-c4f3b40124b1" alt="Screenshot of GOV.UK Design System homepage on iOS, showing the VoiceOver focus ring around a skip link that is not visible" width="501px"/>

## Environment (where applicable)
- Operating system: iOS, iPadOS
- Browser: Safari
- Browser version: Anything less than a year old
- GOV.UK Frontend Version: 4.7.0

**Comment by querkmachine** [comment]
I'm not sure if we have a viable way of fixing this. VoiceOver focus is not equivalent to keyboard focus, it exists outside of the browser scope, and does not seem to emit any events to the page that we could potentially hook on to.

This same issue seems to be present on every website that uses the hidden-until-focused skiplink pattern (BBC.co.uk, for example), so my _assumption_ is that VO users will be pretty used to this happening already.

@selfthinker Given the above, do you think we should close this, given it's unactionable?

**Comment by selfthinker** [comment]
@querkmachine, I generally agree.
However, I'm confused by a couple of things...

> The skip link does become visible on screen reader focus for other screen readers.

That is incorrect. The skip link does not become visible for any other screen reader. It would surprise me if it did.
Unless this is not about the screen reader focus.

> From the top of the page, press 'tab' once

You can only press 'tab' with a keyboard. But the issue only mentions the screen reader focus. That's confusing because when pressing 'tab' on a keyboard, that would also move the keyboard focus.
(I haven't tested this with iOS and a keyboard, as I have no means to do that at the moment.)

Is this issue really talking about screen reader focus or about keyboard focus?

**Comment by querkmachine** [comment]
It seems that iOS with a hardware keyboard attached will show **and** announce the skiplink if navigated to via <kbd>Tab</kbd>[^1], but will announce it **without** showing it if navigated to via gestures (the same as without a keyboard attached).

This makes sense to me, as when the <kbd>Tab</kbd> key was pressed, VoiceOver moved its own focus to match wherever keyboard focus has moved to, presumably to highlight what was now being announced.

There was an interesting bit of behaviour where the skiplink remained visible without focus when tabbing in reverse, as it's the last thing on the page to have focus before keyboard focus entered the user interface instead. I don't think that requires any action, however. 

[^1]: Doing this requires going into the settings and enabling Full Keyboard Access. By default, links cannot be accessed by tabbing on iOS Safari, only form inputs.

**Comment by selfthinker** [comment]
Thanks for double checking @querkmachine.
Based on our combined observations I assume that the original description was just a bit unclear.

I'm happy to close this.
No screen reader that I am aware of would make the skip link visible on screen reader focus, so that is expected behaviour.
And all of them make the link correctly visible on keyboard focus.
